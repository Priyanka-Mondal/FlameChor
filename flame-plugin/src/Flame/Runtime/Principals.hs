{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE PostfixOperators #-}

module Flame.Runtime.Principals
       ( Prin(..)
       , ActsForProof(..)
       , DelClosure
       , voiceOf, eyeOf
       , normalize, actsFor, computeDelClosures
       )
where

import Data.Data
import Control.Arrow  ((***))
import Data.Maybe     (mapMaybe)
import Data.Graph 
import Data.List
import Data.Either
import Data.Data
import Data.Text      (Text)

-- | The principal data type 
data Prin =
  Top
  | Bot
  | Name Text
  | Conj  Prin Prin 
  | Disj  Prin Prin
  | Conf  Prin
  | Integ Prin
  deriving (Ord, Read, Eq, Show, Data, Typeable)

public        = Conf Bot
trusted       = Integ Top
publicTrusted = Conj public trusted           

voiceOf :: Prin -> Prin
voiceOf p = let (N conf _) = normPrin p in
              reify $ N (J [M [B]]) conf

eyeOf :: Prin -> Prin
eyeOf p = let (N _ integ) = normPrin p in
              reify $ N integ (J [M [B]])

-------------------------- Acts for ----------------------------

-- | Proof trees for acts for derivations
data ActsForProof =                    
      AFTop
    | AFBot
    | AFRefl 
    | AFDel (Prin, Prin) -- NB: this implicitly uses transitivity on given dels
    | AFConj [ActsForProof] -- for each RHS q, proof that there is a LHS p s.t. p > q 
    | AFDisj [ActsForProof] -- for each RHS q, proof that there is a LHS p s.t. p > q 
  deriving (Read, Eq, Show, Data, Typeable)

-- | The transitive closure of each principal's superiors.
type DelClosure = [(JNorm, [JNorm])]

-- | Find a proof of an acts-for relationship, if one exists.
actsFor :: [(Prin, Prin)]-- ^ [D]elegations
           -> (Prin , Prin)
           -> Maybe ActsForProof
actsFor delegations (Top, q) = Just AFTop
actsFor delegations (p , Bot) = Just AFBot
actsFor delegations (p,q)
  | p == q    = Just AFRefl
  | otherwise = --pprTrace "actsFor" (ppr (p,q)) $
        let p' = normPrin p in
        let q' = normPrin q in do
          confPf <- confActsFor (conf p', conf q')
          integPf <- integActsFor (integ p', integ q')
          Just $ AFConj [confPf, integPf]
  where
    top :: Norm
    top = N (J [M [T]]) (J [M [T]])
    bot :: Norm
    bot = N (J [M [B]]) (J [M [B]])

    (confClosure, integClosure) = computeDelClosures delegations

    confActsFor :: (JNorm, JNorm) -> Maybe ActsForProof
    confActsFor = actsForJ True confClosure 
    integActsFor :: (JNorm, JNorm) -> Maybe ActsForProof
    integActsFor = actsForJ False integClosure

actsForJ :: Bool ->
            DelClosure ->
            (JNorm, JNorm) ->
            Maybe ActsForProof
actsForJ isConf delClosure (p,q) 
  | p == top  = Just AFTop
  | q == bot  = Just AFBot
  | p == q    = Just AFRefl
  | otherwise = --pprTrace "actsForJ" (ppr (p,q)) $
                case conjProofs of
                  Just [conj] -> Just conj
                  _      -> AFConj <$> conjProofs 
  where
    top :: JNorm
    top = J [M [T]]
    bot :: JNorm
    bot = J [M [B]] 
    -- for every join comp on rhs, find sup on lhs
    (J pms, J qms) = (p,q)
    conjProofs :: Maybe [ActsForProof]
    conjProofs = sequence $ map (\qm ->
                                  case mapMaybe (\pm ->
                                                  actsForM isConf delClosure (pm,qm))
                                                pms
                                  of
                                    (pf:pfs) ->
                                      --pprTrace "found proof" (ppr pf) $
                                      Just pf
                                    [] -> Nothing
                                )
                                qms

actsForM :: Bool ->
            DelClosure ->
            (MNorm, MNorm) ->
            Maybe ActsForProof
actsForM isConf delClosure (p,q) 
  | p == top  = Just AFTop
  | q == bot  = Just AFBot
  | p == q    = Just AFRefl
  | otherwise = case disjProofs of
                  Just [disj] -> Just disj
                  _      -> AFDisj <$> disjProofs 
  where
    top :: MNorm
    top = M [T]
    bot :: MNorm
    bot = M [B] 
    -- for every meet comp on lhs, find inf on rhs
    (M pbs, M qbs) = (p,q)
    disjProofs :: Maybe [ActsForProof]
    disjProofs = sequence $ map (\pb ->
                                  case mapMaybe (\qb ->
                                                  actsForB isConf delClosure (pb,qb))
                                                qbs
                                  of
                                    (pf:pfs) -> Just pf
                                    [] -> Nothing
                                )
                                pbs

-- IDEA for transitivity.  If all given dels are expressed "primitively",
-- then transitivity can be exploited as simple reachability via given dels.
actsForB :: Bool ->
            DelClosure ->
            (Base, Base) ->
            Maybe ActsForProof
actsForB isConf delClosure (p,q) 
  | p == top = Just AFTop
  | q == bot = Just AFBot
  | p == q  = Just AFRefl
  | otherwise = --pprTrace "actsForB" (ppr (p,q)) $
    case find (== J [M [p]]) (superiors $ J [M [q]]) of
      Just del -> if isConf then
                     Just $ AFDel (Conf (reifyBase p), Conf (reifyBase q))
                  else
                     Just $ AFDel (Integ (reifyBase p), Integ (reifyBase q))
      _ -> Nothing
  where
    top :: Base
    top = T
    bot :: Base
    bot = B  
    superiors :: JNorm -> [JNorm]
    superiors q = case find ((== q) . fst) delClosure of
                    Just (q, sups) -> sups
                    _ -> []

--------------------- Principal normalization ------------------------------------

-- | Normalize a principal
normalize :: Prin -> Prin
normalize p = reify $ normPrin p 

-- | Base principals for FLAM Join Normal Form
-- Each Base principal is a primitive principal.
data Base =
  P Text -- ^ Primitive principal
  | B   -- ^ Bottom
  | T   -- ^ Top
  deriving (Eq, Show, Ord)

-- | Meet principals for FLAM Join Normal Form
-- Each MNorm is a meet of base principals.
newtype MNorm = M { unM :: [Base]}
  deriving (Eq, Ord, Show)

-- | Join principals for FLAM Join Normal Form
-- Each JNorm is a join of meets.
newtype JNorm = J { unJ :: [MNorm]}
  deriving (Eq, Ord, Show)

-- | A principal in FLAM Join Normal Form
-- Each Norm is a join of two principals: the confidentiality projection and the
-- integrity projection, each in JNF.
data Norm = N {conf :: JNorm, integ :: JNorm}
  deriving (Eq, Ord, Show)

-- | Convert a 'Prin' to a 'JNorm' term
normPrin :: Prin -> Norm
normPrin Top        = N (J [M [T]]) (J [M [T]])
normPrin Bot        = N (J [M [B]]) (J [M [B]])
normPrin (Name s)   = N (J [M [P s]]) (J [M [P s]])
normPrin (Conf p)   = N (jnormPrin True p) (J [M [B]]) 
normPrin (Integ p)  = N (J [M [B]]) (jnormPrin False p)
normPrin (Conj p q) = mergeNormJoin (normPrin p) (normPrin q)
normPrin (Disj p q) = mergeNormMeet (normPrin p) (normPrin q)

reify :: Norm -> Prin
reify (N c i) = Conj (Conf $ reifyJ c) (Integ $ reifyJ i)
  where
 reifyJ :: JNorm -> Prin
 reifyJ (J [m])    = reifyM m
 reifyJ (J (m:ms)) = Conj (reifyM m) (reifyJ (J ms))

 reifyM :: MNorm -> Prin
 reifyM (M [b])    = reifyBase b
 reifyM (M (b:bs)) = Disj (reifyBase b) (reifyM (M bs))

 -- | Convert a Base principal to a Prin
reifyBase :: Base -> Prin
reifyBase B = Bot
reifyBase T = Top
reifyBase (P s) = Name s

mergeWith :: (a -> a -> Either a a) -> [a] -> [a]
mergeWith _ []      = []
mergeWith op (f:fs) = case partitionEithers $ map (`op` f) fs of
                        ([],_)              -> f : mergeWith op fs
                        (updated,untouched) -> mergeWith op (updated ++ untouched)

-- | Merge two symbols of a Meet term
--
-- Performs the following rewrites:
--
-- @
-- ⊤ ∨ x    ==>  x
-- x ∨ ⊤    ==>  x
-- x ∨ ⊥    ==>  0
-- ⊥ ∨ x    ==>  0
-- x ∨ x    ==>  x
-- @
mergeB :: Base -> Base -> Either Base Base
mergeB T r = Left r       -- ⊤ ∨ x ==> x
mergeB l T = Left l       -- x ∨ ⊤ ==> x
mergeB B _ = Left B       -- ⊥ ∨ x ==> ⊥
mergeB _ B = Left B       -- x ∨ ⊥ ==> ⊥
mergeB l r                -- y ∨ y ==> y
  | l == r = Left l
mergeB l _ = Right l

-- | Merge two components of a Join term
--
-- Performs the following rewrites:
--
-- @
-- ⊤ ∧ x       ==>  ⊤ 
-- x ∧ ⊤       ==>  ⊤
-- ⊥ ∧ x       ==>  x
-- x ∧ ⊥       ==>  x
-- x ∧ (x ∨ y) ==>  x  (Absorbtion)
-- (x ∨ y) ∧ x ==>  x  (Absorbtion)
-- @
mergeM :: MNorm -> MNorm -> Either MNorm MNorm
mergeM (M [T]) _ = Left (M [T])                   -- ⊤ ∧ x       ==>  ⊤ 
mergeM _ (M [T]) = Left (M [T])                   -- x ∧ ⊤       ==>  ⊤ 
mergeM (M (B:_)) r = Left r                       -- ⊥ ∧ x       ==>  x 
mergeM l (M (B:_)) = Left l                       -- x ∧ ⊥       ==>  x
mergeM (M [l]) (M rs) | elem l rs = Left (M [l])  -- x ∧ (x ∨ y) ==>  x
mergeM (M ls) (M [r]) | elem r ls = Left (M [r])  -- (x ∨ y) ∧ x  ==>  x
mergeM l r | l == r = Left l                      -- y ∧ y    ==>  y
mergeM l _ = Right l

zeroM :: MNorm -> Bool
zeroM (M (B:_)) = True
zeroM _         = False

mkNonEmpty :: JNorm -> JNorm 
mkNonEmpty (J []) = J [M [B]]
mkNonEmpty s      = s

-- | Simplifies SOP terms using
--
-- * 'mergeM'
-- * 'mergeB'
simplifyJNorm :: JNorm -> JNorm 
simplifyJNorm = repeatF go
  where
    go = mkNonEmpty
       . J
       . sort . filter (not . zeroM)
       . mergeWith mergeM
       . map (M . sort . mergeWith mergeB . unM)
       . unJ

    repeatF f x =
      let x' = f x
      in  if x' == x
             then x
             else repeatF f x'
{-# INLINEABLE simplifyJNorm #-}

-- | Merge two JNorm terms by join
mergeJNormJoin :: JNorm -> JNorm -> JNorm 
mergeJNormJoin (J ms1) (J ms2) = simplifyJNorm $ J (ms1 ++ ms2)
{-# INLINEABLE mergeJNormJoin #-}

-- | Merge two JNorm terms by meet
mergeJNormMeet :: JNorm -> JNorm -> JNorm
mergeJNormMeet (J ms1) (J ms2)
  = simplifyJNorm
  . J
  $ concatMap (zipWith (\p1 p2 -> M (unM p1 ++ unM p2)) ms1 . repeat) ms2
{-# INLINEABLE mergeJNormMeet #-}

-- | Merge two Norm terms by join
mergeNormJoin :: Norm -> Norm -> Norm 
mergeNormJoin (N c1 i1) (N c2 i2) = N (mergeJNormJoin c1 c2) (mergeJNormJoin i1 i2)
{-# INLINEABLE mergeNormJoin #-}

-- | Merge two Norm terms by meet
mergeNormMeet :: Norm -> Norm -> Norm
mergeNormMeet (N c1 i1) (N c2 i2) = N (mergeJNormMeet c1 c2) (mergeJNormMeet i1 i2)
{-# INLINEABLE mergeNormMeet #-}

-- | Convert a 'Prin' to a 'JNorm' term
-- isConf indicates whether we are normalizing the conf component
jnormPrin :: Bool -> Prin -> JNorm
jnormPrin isConf Top = J [M [T]]
jnormPrin isConf Bot = J [M [B]]
jnormPrin isConf (Name s) = J [M [P s]]
jnormPrin isConf (Conf p) = 
  if isConf then jnormPrin isConf p else J [M [B]]
jnormPrin isConf (Integ p) = 
  if isConf then J [M [B]] else jnormPrin isConf p
jnormPrin isConf (Conj p q) =
  mergeJNormJoin (jnormPrin isConf p) (jnormPrin isConf q)
jnormPrin isConf (Disj p q) =
  mergeJNormMeet (jnormPrin isConf p) (jnormPrin isConf q)


--------------------- Delegation closures ------------------------------------

{-
    - expand given constraints to "base form": conf or integ, no RHS conj, no LHS disj
    - for each conjunction on the LHS, add a pseudo-node to the graph that is
        reachable from each conjunct and from which the intersection of the
        superiors of each conjunct are reachable.
    - for each disjunction on the RHS, add a pseudo-node to the graph that
        reaches the disjunction and is reachable from the intersection of
        the inferiors of each disjunct.
    - fixpoint?
-}
computeDelClosures :: [(Prin,Prin)] -> (DelClosure, DelClosure) 
computeDelClosures dels = let (confGivens, integGivens) = expandGivens dels in
                           (givenClosure confGivens, givenClosure integGivens)

-- | Expand given delegations to "base form".  Base-form delegations are split
--   into confidentiality delegations and integrity delegations, have no meets
--   on the left-hand side, and no joins on the right-hand side.
--   Thus, base-form delegations all have the form: (b ∧ b ...) ≽ (b ∨ b ...)
expandGivens :: [(Prin,Prin)]
             -> ([(JNorm,JNorm)], [(JNorm,JNorm)])
expandGivens givens = foldl
                      (\(cacc, iacc) given ->
                        case given of
                          (N supJC supJI, N (J infMCs) (J infMIs)) -> 
                            ([(supC, J [infC]) | supC <- cartProd supJC, infC <- infMCs] ++ cacc,
                             [(supI, (J [infI])) | supI <- cartProd supJI, infI <- infMIs] ++ iacc)
                      )
                      ([] :: [(JNorm, JNorm)], [] :: [(JNorm, JNorm)])
                      [(normPrin p, normPrin q) | (p, q) <- givens]
  
-- | Compute closure of given delegations.  Expects delegations to be in base form.
givenClosure :: [(JNorm,JNorm)] -> DelClosure
givenClosure givens =

  [(inferior, superiors) | (inferior, _, superiors) <- fixpoint initialEdges]
    
  where
    top = (J [M [T]])
    bot = (J [M [B]])
    baseSeqToJ seq = J [M seq]
    {-
      For principals in a given in base form, 
        (b ∧ b ...) ≽ (b ∨ b ...)
      we want a node for each subsequence of conjuncts and disjuncts
    -}
    structJoinEdges :: JNorm -> [(JNorm, JNorm, [JNorm])]
    structJoinEdges (J []) = [] 
    structJoinEdges (J seq) = 
      [(J inf, J inf, [J seq]) | inf <- subsequencesOfSize (length seq - 1) seq]
      ++ concat [structJoinEdges (J inf) | inf <- subsequencesOfSize (length seq - 1) seq]

    structMeetEdges :: JNorm -> [(JNorm, JNorm, [JNorm])]
    structMeetEdges (J [M []]) = [] 
    structMeetEdges (J [M seq]) = 
      [(baseSeqToJ seq, baseSeqToJ seq, map baseSeqToJ $ subsequencesOfSize (length seq - 1) seq)]
      ++ concat [structMeetEdges (J [M sup]) | sup <- subsequencesOfSize (length seq - 1) seq]

    initialEdges :: [(JNorm, JNorm, [JNorm])]
    initialEdges = [(inf, inf, union (union (nub [gsup | (gsup, ginf) <- givens, ginf == inf])
                                            $ concat [jsups | (jinf, _, jsups) <- structJoinEdges inf, jinf == inf])
                                     $ concat [msups | (minf, _, msups) <- structJoinEdges inf, minf == inf])
                    | inf <- principals]

    principals :: [JNorm]
    principals = [top, bot] ++ (nub $ concat [(map J $ concat [subsequencesOfSize i psC | i <- [1..length psC]]) ++
                                              (map baseSeqToJ $ concat [subsequencesOfSize i qs | i <- [1..length qs]])
                                             | (J psC, J [M qs]) <- givens])

    fixpoint edges = let (graph, vtxToEdges, prinToVtx) = graphFromEdges edges in
                     let vtxToPrin v = let (x, _, _) = vtxToEdges v in x in
                     let newEdges = [(vtxToPrin inf, vtxToPrin inf, 
                                                        (map vtxToPrin $ reachable graph inf) ++
                                                        computeStructEdges (graph, vtxToEdges, prinToVtx) inf)
                                    | inf <- vertices graph] in
                     if edges == newEdges then
                       newEdges 
                     else
                       fixpoint newEdges

    computeStructEdges (graph, vtxToEdges, prinToVtx) vtx = []

subsequencesOfSize :: Int -> [a] -> [[a]]
subsequencesOfSize n xs = let l = length xs
                          in if n>l then [] else subsequencesBySize xs !! (l-n)
 where
   subsequencesBySize [] = [[[]]]
   subsequencesBySize (x:xs) = let next = subsequencesBySize xs
                             in zipWith (++) ([]:next) (map (map (x:)) next ++ [[]])

-- | Cartesian product of a JNorm principal
--   This converts from a join of meets to a meet of joins.
cartProd :: JNorm -> [JNorm]
cartProd (J ms) = [J $ map mkM ps | ps <- sequence [bs | (M bs) <- ms]]
  where mkM p = M [p]
