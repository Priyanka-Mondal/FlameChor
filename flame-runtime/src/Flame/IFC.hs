{-# LANGUAGE GADTs, PolyKinds, KindSignatures, MultiParamTypeClasses,
    DataKinds, RankNTypes, FlexibleInstances, FlexibleContexts, TypeFamilies #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# OPTIONS_GHC -fplugin Flame.Solver -fobject-code #-}

module Flame.IFC
       ( Labeled(label, unlabel,relabel), Lbl
       , IFC(lift, apply, bind, runIFC, protect, use, reprotect, ffmap, fjoin)
       , FLA(..), FLAC, FLACT, NMIF(..), NM, NMT, runFLAC, runNM 
       , (:≽), (:=>=), (:⊑), (:<:)          -- Delegation types
       , Def(..), (≽), (=>=), (⊑), (<:)     -- Delegation constructors
       )
  where

import Flame.TCB.IFC
import Flame.TCB.Assume
