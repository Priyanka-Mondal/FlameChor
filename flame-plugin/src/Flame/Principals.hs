{-# LANGUAGE GADTs, PolyKinds, DataKinds, RankNTypes #-}
{-# LANGUAGE TypeOperators, PostfixOperators #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE DeriveDataTypeable #-}

module Flame.Principals
       ( KPrin (..)
       , SPrin (..)
       , C, I, type (^->), type (^→), type (^<-), type (^←)
       , N, Public, Secret, Trusted, Untrusted, PT, SU 
       , DPrin (st, dyn), Ex(..) 
       , withPrin, promote
       , type (/\), type (∧), type (\/), type (∨) 
       , type (⊔), type (⊓)
       , type (≽), type (>=), type (⊑), type (<:), type (===)
       , type (∇), Voice, Δ, Eye, (∇), δ 
       , (⊤), top, (⊥), bot
       , public, trusted, publicTrusted, secret, untrusted, secretUntrusted
       , (^->), (^→), (^<-), (^←)
       , (/\), (∧), (\/), (∨), (⊔), (⊓)
       , (*->), (*→), (*<-), (*←), (*/\), (*∧), (*\/), (*∨), (*⊔), (*⊓), (*∇)  
       , (<=>) -- XXX TODO: should not export to non-TCB code.
       )
where

import Data.Constraint hiding (top)
import GHC.TypeLits
import Data.Proxy (Proxy(..))
import Data.Type.Bool
import Data.Data
import Data.Text  (pack, unpack)

import Flame.Runtime.Principals

{-| The principal kind. Type-level FLAM principals |-}
data KPrin =
  KTop
  | KBot
  | KName  Symbol 
  | KConj  KPrin KPrin 
  | KDisj  KPrin KPrin
  | KConf  KPrin
  | KInteg KPrin
  | KVoice KPrin
  | KEye   KPrin

{-| 
Singleton GADT for KPrin. This GADT associates a KPrin (which is an
uninstantiable kind) with a single runtime value, providing a link
between runtime principals and type-level principals.  NB: we employ
an explicit singleton pattern here b/c the singletons package cannot
promote strings to symbols.
|-}
data SPrin :: KPrin -> * where
  STop   :: SPrin KTop
  SBot   :: SPrin KBot
  SName  :: forall (n :: Symbol). { sym :: Proxy n } -> SPrin (KName n)
  SConj  :: !(SPrin p) -> !(SPrin q) -> SPrin (KConj p q)
  SDisj  :: !(SPrin p) -> !(SPrin q) -> SPrin (KDisj p q)
  SConf  :: !(SPrin p) -> SPrin (KConf p)
  SInteg :: !(SPrin p) -> SPrin (KInteg p)
  SVoice :: !(SPrin p) -> SPrin (KVoice p)
  SEye   :: !(SPrin p) -> SPrin (KEye   p)

deriving instance Show (SPrin p)
deriving instance Eq (SPrin p)

{- Some notation help -}
type C p      = KConf p
type (^→) p  = KConf p
type (^->) p  = KConf p
type I p      = KInteg p
type (^←) p  = KInteg p
type (^<-) p  = KInteg p
type Voice p  = KVoice p
type Eye p    = KEye p
type N s      = KName s

type (/\) p q = KConj p q
type p ∧ q  = KConj p q
infixl 7 /\
infixl 7 ∧

type (\/) p q = KDisj p q
type (∨)  p q = KDisj p q
infixl 7 ∨

type (∇) p = Voice p
type (Δ) p = Eye p

-- TODO: convert flow join/meet to type families so that
--       resulting types are more normalized
type (⊔) p q  = (C p ∧ C q) ∧ (I p ∨ I q)
infixl 6 ⊔
type (⊓) p q  = (C p ∨ C q) ∧ (I p ∧ I q)
infixl 6 ⊓

type (⊤) = KTop
type (⊥) = KBot
type Public   = C KBot
type Trusted  = I KTop
type PT       = Public ∧ Trusted 

type Secret = C KTop
type Untrusted  = I KBot
type SU       = Secret ∧ Untrusted

(*->) p   = SConf p
(*→)  p   = SConf p

(*<-) p   = SInteg p
(*←) p   = SInteg p

(*∇)  p = SVoice p

(*/\) p q  = SConj p q
(*∧)  p q  = SConj p q

(*\/) p q  = SDisj p q
(*∨)  p q  = SDisj p q

(*⊔)  p q  = ((p*→) *∧ (q*→)) *∧ ((p*←) *∨ (q*←))
(*⊓)  p q  = ((p*→) *∨ (q*→)) *∧ ((p*←) *∧ (q*←))

public :: SPrin Public
public = SConf SBot
trusted  :: SPrin Trusted
trusted  = SInteg STop
publicTrusted :: SPrin PT
publicTrusted = public *∧ trusted

secret :: SPrin Secret
secret = (SConf STop)
untrusted  :: SPrin Untrusted
untrusted  = (SInteg SBot)
secretUntrusted :: SPrin SU
secretUntrusted = secret *∧ untrusted

{- Existential wrapper -}
data Ex (p :: k -> *) where
  Ex :: p i -> Ex p

{- Promote runtime principals to existentially-wrapped principal types -}
promoteS :: Prin -> Ex SPrin
promoteS p =
  case p of
    Top         ->  Ex STop
    Bot         ->  Ex SBot
    (Name str)  ->  case someSymbolVal (unpack str) of
                      SomeSymbol n -> Ex (SName n)
    (Conj p q)  ->  case promoteS p of
                      Ex p' -> case promoteS q of
                                 Ex q' -> Ex (SConj p' q')
    (Disj p q)  ->  case promoteS p of
                      Ex p' -> case promoteS q of
                                 Ex q' -> Ex (SDisj p' q')
    (Conf p)    ->  case promoteS p of Ex p' -> Ex (SConf p')
    (Integ p)   ->  case promoteS p of Ex p' -> Ex (SInteg p')

{- Bind runtime principal to type -}
withPrin :: Prin -> (forall p . DPrin p -> a) -> a
withPrin p f = case promote p of
                 Ex p' -> f p'

promote :: Prin -> Ex DPrin
promote p = case promoteS p of
              Ex p' -> Ex (p <=> p')


{-| DPrin p is a record associating a dynamic principal with a
singleton principal. This construction is necessary since not all
principal names may be known at compile-time, yet we still 
want to construct a type-level representation for runtime principals.

Restricting <=> and UnsafeAssoc to trusted code ensures only withPrin
can associate runtime principals wth singleton principal types.
|-}
data DPrin p = UnsafeAssoc { dyn :: Prin, st :: !(SPrin p) } 
dynamic = dyn
static = st 
(<=>) :: Prin -> SPrin p -> DPrin p
p <=> sp = UnsafeAssoc p sp

{-| Notation helpers for DPrin values |-} 
(⊤) :: DPrin (⊤)
(⊤) = Top <=> STop
top = (⊤)

(⊥) :: DPrin (⊥)
(⊥) = Bot <=> SBot
bot = (⊥)

(^→) :: DPrin p  -> DPrin (C p)
(^→) p  = Conf (dyn p) <=> SConf (st p)
(^->) = (^→)

(^←) :: DPrin p  -> DPrin (I p)
(^←) p = Integ (dyn p) <=> SInteg (st p)
(^<-) = (^←)

(∧) :: DPrin p -> DPrin q -> DPrin (p ∧ q) 
(∧) p q = Conj (dyn p) (dyn q) <=> SConj (st p) (st q)
(/\) = (∧)

(∨) :: DPrin p -> DPrin q -> DPrin (p ∨ q) 
(∨)  p q  = Disj (dyn p) (dyn q) <=> SDisj (st p) (st q)
(\/) = (∨)

(⊔) :: DPrin p -> DPrin q -> DPrin (p ⊔ q) 
(⊔)  p q  = (Conj (Conf (Conj (dyn p) (dyn q)))
             (Integ (Disj (dyn p) (dyn q)))) <=> ((st p) *⊔ (st q))
join = (⊔)

(⊓) :: DPrin p -> DPrin q -> DPrin (p ⊓ q) 
(⊓) p q  = (Conj (Conf (Disj (dyn p) (dyn q)))
            (Integ (Conj (dyn p) (dyn q)))) <=> ((st p) *⊓ (st q))
meet = (⊓)

(∇) :: DPrin p -> DPrin ((∇) p)
(∇) p = voiceOf (dyn p) <=> SVoice (st p)
        
δ :: DPrin p -> DPrin (Δ p)
δ p = eyeOf (dyn p) <=> SEye (st p)

{-| Actsfor constraints. This type family is closed: only the Flame solver is capable of resolving these constraints.  |-}
type family (≽) (p :: KPrin) (q :: KPrin) :: Constraint where
infixl 4 ≽

{- Type synonyms for acts for constraints. As in FLAM, the safe
information flow relation is defined in terms of the acts-for
relation. -}
type (>=) (p :: KPrin) (q :: KPrin) = (p ≽ q) 
infixl 4 >=
type (⊑) (p :: KPrin) (q :: KPrin) = (C q ≽ C p , I p ≽ I q) 
infixl 4 ⊑
type (<:) (p :: KPrin) (q :: KPrin) = (p ⊑ q)
infixl 4 <:
type (===) (p :: KPrin) (q :: KPrin) = (p ≽ q, q ≽ p)
infixl 4 ===
