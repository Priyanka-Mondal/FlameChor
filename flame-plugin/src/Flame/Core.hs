module Flame.Core
       ( module Flame.IFC
       , module Flame.Principals
       , module Flame.Runtime.Principals
       , module Flame.Do
       )
  where

import Flame.IFC
import Flame.Principals
import Flame.Runtime.Principals
import Flame.Do

