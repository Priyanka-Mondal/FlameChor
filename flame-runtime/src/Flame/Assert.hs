{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE PostfixOperators #-}
{-# LANGUAGE FlexibleContexts #-}
{-# OPTIONS_GHC -fplugin Flame.Solver -fobject-code #-}
module Flame.Assert
where
  
import Flame.Principals
assertEq :: (l === l') => SPrin l -> SPrin l' -> ()
assertEq l l' = ()

assertActsFor :: (p ≽ q) => SPrin p -> SPrin q -> ()
assertActsFor p q = ()

assertFlowsTo :: (l ⊑ l') => SPrin l -> SPrin l' -> ()
assertFlowsTo l l' = ()
