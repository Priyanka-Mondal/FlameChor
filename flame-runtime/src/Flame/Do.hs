{-# LANGUAGE TemplateHaskell #-}
{-# OPTIONS_GHC -fplugin Flame.Solver -fobject-code #-}

----------------------------------------------------------------------------
---- Based on Language.Haskell.IndexedDo by Fumiaki Kinoshita
----------------------------------------------------------------------------
module Flame.Do (flame) where

import Language.Haskell.Meta
import Language.Haskell.TH
import Language.Haskell.TH.Quote
import qualified Language.Haskell.Exts.Syntax as Hs
import qualified Language.Haskell.Exts.SrcLoc as Hs
import Language.Haskell.Exts.Extension as Ext
import Language.Haskell.Exts.Parser (ParseMode(..), ParseResult(..), parseExpWithMode)

import Flame.IFC

myDefaultParseMode2 :: ParseMode
myDefaultParseMode2 = ParseMode
  {parseFilename = []
  ,baseLanguage = Haskell2010
  ,extensions = map EnableExtension myDefaultExtensions2
  ,ignoreLinePragmas = False
  ,ignoreLanguagePragmas = False
  ,fixities = Nothing
  ,ignoreFunctionArity = False
  }

myDefaultExtensions2 :: [KnownExtension]
myDefaultExtensions2 = [Ext.PostfixOperators
                      ,Ext.QuasiQuotes
                      ,Ext.UnicodeSyntax
                      ,Ext.PatternSignatures
                      ,Ext.MagicHash
                      ,Ext.ForeignFunctionInterface
                      ,Ext.TemplateHaskell
                      ,Ext.RankNTypes
                      ,Ext.MultiParamTypeClasses
                      ,Ext.RecursiveDo
                      ,Ext.TypeApplications]
  
myParseExp :: String -> Either String Exp
myParseExp = either Left (Right . toExp) . myParseHsExp

myParseHsExp :: String -> Either String (Hs.Exp Hs.SrcSpanInfo)
myParseHsExp = parseResultToEither . parseExpWithMode myDefaultParseMode2

flame :: QuasiQuoter
flame = QuasiQuoter { quoteExp = \str -> case myParseExp str of
        Right (DoE Nothing ss) -> return (go ss)
        Right _ -> fail "Expecting do notation"
        Left err -> fail (show err)
        , quotePat = undefined
        , quoteType = \str -> case parseType str of
        Right t -> return t
        Left err -> fail (show err)
        , quoteDec = undefined
        } where
    go [NoBindS e] = e
    go (BindS p e : ss) = VarE 'use `AppE` e `AppE` LamE [p] (go ss)
    go (NoBindS e : ss) = VarE 'apply `AppE` e `AppE` LamE [WildP] (go ss)
    go (LetS ds : ss) = LetE ds (go ss)
