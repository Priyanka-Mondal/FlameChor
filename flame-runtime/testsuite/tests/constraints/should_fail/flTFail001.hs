{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE PostfixOperators #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE RankNTypes #-}
{-# OPTIONS_GHC -fplugin Flame.Solver #-}
module ShouldFail where

import Flame.Assert
import Flame.Principals

neg_flTConf ::  SPrin p -> ()
neg_flTConf p = assertFlowsTo ((p*→) *∧ (SBot*←)) p
