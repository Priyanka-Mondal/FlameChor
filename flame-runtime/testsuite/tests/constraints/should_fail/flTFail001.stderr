flTFail001.hs:14:17:
    Could not deduce: (C p ≽ C ('KConj ('KConf p) ('KInteg 'KBot)),
                        I ('KConj ('KConf p) ('KInteg 'KBot)) ≽ I p)
       arising from a use of ‘assertFlowsTo’
    In the expression: assertFlowsTo ((p *→) *∧ (SBot *←)) p
     In an equation for ‘neg_flTConf’:
         neg_flTConf p = assertFlowsTo ((p *→) *∧ (SBot *←)) p
    Relevant bindings include
       p :: SPrin p (bound at flTFail001.hs:14:13)
       neg_flTConf :: SPrin p -> () (bound at flTFail001.hs:14:1)

